File: leetcode/algorithms/202_happy_number/main.py
class HappyNumber:
    """
    # Hash Map
    #
    # Complexities:
    #   N - `n`
    #   - Time Complexity: O(logᴺ)
    #   - Space Complexity: O(logᴺ)
    """
    def isHappy(self, n: int) -> bool:
        if n <= 1:
            return True

        numDict: dict = {}

        while n > 1:
            nString = str(n)

            if numDict.get(nString) is True:
                return False

            numDict[nString] = True
            n = 0

            for c in nString:
                n += int(c) ** 2

            if n is 1:
                return True

        return False


    # Solution
    """
    # Solution 1
    #
    # Hash Set
    #
    # Complexities:
    #   N - `n`
    #   - Time Complexity: O(logᴺ)
    #   - Space Complexity: O(logᴺ)
    """
    def solution1(self, n: int) -> bool:
        def get_next(number: int) -> int:
            total_sum = 0

            while number > 0:
                number, digit = divmod(number, 10)
                total_sum += digit ** 2

            return total_sum
        
        seen = set()
        
        while n != 1 and n not in seen:
            seen.add(n)
            n = get_next(n)
            
        return n == 1


    """
    # Solution 2
    #
    # Floyd's Cycle-Finding Algorithm
    #
    # Complexities:
    #   N - `n`
    #   - Time Complexity: O(logᴺ)
    #   - Space Complexity: O(1)
    """
    def solution2(self, n: int) -> bool:
        def get_next(number: int) -> int:
            total_sum = 0

            while number > 0:
                number, digit = divmod(number, 10)
                total_sum += digit ** 2

            return total_sum
        
        slow = n
        fast = get_next(n)
        
        while fast != 1 and slow != fast:
            slow = get_next(slow)
            fast = get_next(get_next(fast))
            
        return fast == 1

File: leetcode/algorithms/202_happy_number/main.cpp
#include <unordered_set>
using namespace std;

class HappyNumber {
public:
    /**
     * Floyd's Cycle-Finding Algorithm
     *
     * Complexities:
     *   N - `n`
     *   - Time Complexity: O(logᴺ)
     *   - Space Complexity: O(1)
     */
    int sum(int n) {
        int sum = 0;
        while (n) {
            sum += (n % 10) * (n % 10);
            n /= 10;
        }
        return sum;
    }

    bool isHappy(int n) {
        int fast = n;
        int slow = n;

        while (true) {
            slow = sum(slow);
            fast = sum(fast);
            fast = sum(fast);
            if (fast == 1) {
                break;
            }
            if (slow == fast) {
                return false;
            }
        }

        return true;
    }


    // Solution
    /**
     * Solution 1
     * 
     * Floyd's Cycle-Finding Algorithm
     *
     * Complexities:
     *   N - `n`
     *   - Time Complexity: O(logᴺ)
     *   - Space Complexity: O(1)
     */
    int getNext1(int n) {
        int sum = 0;
        while (n > 0) {
            int digit = n % 10;
            sum += digit * digit;
            n /= 10;
        }
        return sum;
    }

    bool solution1(int n) {
        int slow = n;
        int fast = getNext1(n);

        while (fast != 1 && slow != fast) {
            slow = getNext1(slow);
            fast = getNext1(getNext1(fast));
        }

        return fast == 1;
    }

    /**
     * Solution 2
     * 
     * unordered_set
     *
     * Complexities:
     *   N - `n`
     *   - Time Complexity: O(logᴺ)
     *   - Space Complexity: O(logᴺ)
     */
    int getNext2(int n) {
        int sum = 0;

        while (n > 0) {
            int digit = n % 10;
            sum += digit * digit;
            n /= 10;
        }

        return sum;
    }

    bool solution2(int n) {
        unordered_set<int> seen;

        while (n != 1 && seen.find(n) == seen.end()) {
            seen.insert(n);
            n = getNext2(n);
        }

        return n == 1;
    }
};
